### DISCUSSION TITLE
nvim-dap how to configure target executable?

### DISCUSSION THREAD

**Discussion by jrunestone** [nix-community/nixvim#2512] [General]
| Field        | Description |
|--------------|-------------|
| Plugin       | `nvim-dap`  | 
| Nixpkgs      | `unstable` | 
| Home Manager | `unstable | <!-- [OPTIONAL] `unstable` or specific version -->

- [x] I have read the [FAQ](https://nix-community.github.io/nixvim/user-guide/faq.html) and my bug is not listed there.

## Description

Pretty sure this isn't a bug, but how do I do anything other than hard-coding the path to the target debug executable in the dap configuration section? The [examples over at the nvim-dap repo uses a lua function](https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#ccrust-via-gdb). Can I not pass it like this? I'm not getting any errors but the program doesn't launch. Hard-coding a file path does work.

```nix
programs.nixvim.plugins.dap.configurations.c = [{
  name = "Debug";
  type = "gdb";
  request = "launch";
  program = ''
    function()
      return vim.fn.input('Path to executable: ', vim.fn.getcwd() .. '/', 'file')
    end
  '';
}];
```

Thank you.

**Comment by daniellaing** [comment]
Looking at the docs, this doesn't seem to be an option.
We could change `plugins.dap.configurations.*` to be attrs of anything and allow the user to configure it according to the plugin docs, similar to how `cmp.settings` works.

**Comment by jrunestone** [comment]
OK sounds like a good idea. I thought `program` already worked like that since it's not explicitly specified in the [nixvim config](https://github.com/nix-community/nixvim/blob/main/plugins/dap/dapHelpers.nix#L93) (but `type`, `request`, `name` are). Hard-coding the path doesn't really lend itself to a generic configuration :)

**Comment by jdbosser** [comment]
@jrunestone did you find any workaround?

**Comment by jrunestone** [comment]
No sorry I disabled the plugin. Never really got used to language and debugging support in neovim, I recently switched to zed-editor instead.

**Comment by jdbosser** [comment]
I worked a bit on it, and I got this to work for my rust debugger: 

Here is my `debugging.nix` module for nixvim. It contains a lot of uneccessary parts for this discussion, but I believe that if 
someone from the future stumble into this issue, a complete module will probably be appreciated. 

```nix
{config, helpers, pkgs, ...}:
{
	plugins = {
		dap = {
			enable = true; 
			extensions.dap-ui.enable = true; 
			adapters.servers.codelldb = rec {
				port = "\${port}";
				executable = {
					command = "${pkgs.vscode-extensions.vadimcn.vscode-lldb}/share/vscode/extensions/vadimcn.vscode-lldb/adapter/codelldb";
					args = ["--port" "${port}"];
				}; 
			}; 
			configurations.rust = [{
				type = "codelldb"; 
				name = "Rust Debug"; 
				request = "launch";
				program = helpers.mkRaw "
function()
	vim.fn.jobstart('cargo build') 
	return vim.fn.input('Path to executable: ', vim.fn.getcwd() .. '/', 'file')
end
			"; 
			}

			];

		};
	};

}
```

**Comment by khaneliman** [comment]
The nvim-dap module has yet to be refactored to latest standards, and should get some documentation love when it does.. I also have a fairly robust nvim-dap configuration that I use for debugging multiple different projects for work and side projects https://github.com/khaneliman/khanelivim/blob/fbc004758de53f6b4a67e37e51fa8701b2cc9ccb/packages/khanelivim/plugins/dap.nix if someone needs some more examples of how to configure it.

**Reply by 99linesofcode** [reply]
Hi, thanks for sharing your configuration it has been a super valuable reference for setting up `nvim-dap`. Everything is working nicely but I noticed that [`codelldb` also has support for Cargo](https://github.com/vadimcn/codelldb/blob/master/MANUAL.md#cargo-support) but when I replace `program` with the `cargo` configuration mentioned `nvim-dap` complains that the `program` launch option is missing. 

Seems to me the nixvim plugin passes the configurations as is so I'm a bit at a loss why this is not working. Any pointers on that too? Thanks!
